pkg/sensors/program/coll.go

```diff
@@ -0,0 +1,168 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Tetragon
+
+package program
+
+import (
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/tetragon/pkg/logger"
+)
+
+// The idea of LoadedCollection is to keep loaded programs and maps
+// from ebpf.Collection to be used later for TestLoad* tests to verify
+// loaded programs and maps.
+//
+// We can't just iterate all load programs and maps for 2 reasons:
+//   - we could race with others loading bpf programs
+//   - we get limited name, which we usually cross
+//
+// The process of loading LoadedCollection is following:
+//
+//   - ebpf collection is loaded
+//
+//     coll := NewCollectionWithOptions
+//
+//   - copy all programs/maps from ebpf collection
+//
+//     copyLoadedCollection(lc, coll)
+//
+//   - collection is removed and only 'used' programs and maps stay loaded
+//     so we filter out unused programs/maps with
+//
+//     load.LC = filterLoadedCollection(lc)
+//
+// This was we have only programs/maps realted to our test and with
+// original long names.
+//
+// All this is happening only when program.KeepCollection is set true,
+// so it's enabled only for testing code.
+
+var (
+	KeepCollection bool
+)
+
+type LoadedMap struct {
+	ID ebpf.MapID
+}
+
+type LoadedProgram struct {
+	ID     ebpf.ProgramID
+	MapIDs []ebpf.MapID
+	Type   ebpf.ProgramType
+}
+
+type LoadedCollection struct {
+	Programs map[string]*LoadedProgram
+	Maps     map[string]*LoadedMap
+}
+
+func newLoadedCollection() *LoadedCollection {
+	lc := &LoadedCollection{}
+	lc.Maps = map[string]*LoadedMap{}
+	lc.Programs = map[string]*LoadedProgram{}
+	return lc
+}
+
+func printLoadedCollection(str string, lc *LoadedCollection) {
+	logger.GetLogger().Infof("Programs (%s):", str)
+	for name, lp := range lc.Programs {
+		logger.GetLogger().Infof("%d: %s - %v", lp.ID, name, lp.MapIDs)
+	}
+	logger.GetLogger().Infof("Maps (%s):", str)
+	for name, lm := range lc.Maps {
+		logger.GetLogger().Infof("%d: %s", lm.ID, name)
+	}
+}
+
+func copyLoadedCollection(lc *LoadedCollection, coll *ebpf.Collection) error {
```

Comment:
how about writing this as:
> copyLoadedCollection(coll *ebpf.Collection) (*LoadedCollection, error) 

I think it makes the interface clearer (?)

Reply:
yes, I will put the LoadedCollection as return value, thanks